### ISSUE TITLE
how to merge stacks? Better docs needed on how to create stacks with shared codebases

### ISSUE THREAD

**Issue by breathe** [pulumi/pulumi#3339]
I've read the docs on stacks many many times and I really don't understand sane organizing principals.  Seems like from a mechanism perspective -- stacks are things that are (1) independently deployed (2) can have 'different versions' deployed from same declaration code.

Its a little unclear how to 'add new stacks' in a clean way to an existing multi-stack project.  I want to be able to just create new .yaml files and re-use the node_modules directory from a parent/sibling stack -- but no matter what I do I seem to always have to run pulumi stack init or something similar which generates .yml files (and other files) as a side effect.

I've now got a bunch of stacks that are too granular for my actual needs -- I'm not realizing any sufficient benefits from stack isolation and so would like to merge them all into one stack for simplicity.  How can I do that?

**Comment by brandonkal** [comment]
At least with node, you can create an index file and re-export all exports from each component. The side effect from importing will take care of merging operations.

**Comment by breathe** [comment]
@brandonkal this would require a redeployment of all the resources wouldn't it ...?

I'm trying to manage VM's with pulumi -- and some of these VM's are unfortunately snowflakes -- so I unfortunately can't reproduce the stack in which the vm is deployed by deleting and redeploying ...

I've gotten all the snowflakes migrated into a single stack, and putting everything in that same stack going forward ... I guess I just might have to deal with the few that I can't move ...

**Comment by brandonkal** [comment]
@breathe Yes. I found that out the hard way as well. I ran into lots of performance issues with Pulumi and thus had to break things up to offset the very slow diff speed. I then tried moving code into the main stack and Pulumi now wants to recreate everything.

Pulumi doesn't really let you move resources or refactor code between stacks. Essentially your end goal would be to have everything in one stack. You'd break it up into different modules as needed (for re-use for later projects) and then import your modules in an index.ts file.

FWIW, I'm switching to Deno + Kapp for Kubernetes config generation. By separating the config generation from the deployment tool, I'll see a much faster iteration cycle. I'll use Pulumi for the occasional VM and DNS config.

You could also do some surgical JSON editing on your stack to try to move things around.
